### PR TITLE
cache bodies to avoid reparsing

### PR DESCRIPTION
**Pull request by wonnage**
Attempt to address https://github.com/browserify/resolve/issues/185 by caching JSON.parse results

### CODE REVIEW

File: lib/sync.js [hunk 1]
```diff
@@ -25,6 +25,8 @@ var defaultIsDir = function isDirectory(dir) {
     return stat.isDirectory();
 };
 
+var pkgCache = {};
```
Comment (ljharb) [inline lib/sync.js]: This introduces state into what was previously a pure module. I'm not sure this is a good idea.

**Comment by ljharb** [comment]
Another alternative is to use `require` instead of `readFile`, which will leverage the require cache (altho it'll lose the ability for a package.json file to be modified during the process)

**Comment by wonnage** [comment]
@ljharb I added some instrumentation to the PR and ran the following script in the airbnb repo: 

https://gist.github.com/wonnage/358fdac4d49fda83b2d9796f0212acd8

This resolves a path 10k times, which seems to be approximately on the order of number of modules in our webpack build. 

I also made a small change to hash the contents instead of doing a full string compare in the cache. The result is 6x faster. Complexity probably scales with package.json size (we have a 48kb package json file). 

````
⇒  time node benchmark.js
5200.934117999997
node benchmark.js  6.37s user 0.84s system 100% cpu 7.189 total
⇒  time node benchmark.js
5234.587008999995
node benchmark.js  6.42s user 0.87s system 99% cpu 7.290 total
⇒  time node benchmark.js
5195.409848000002
node benchmark.js  6.35s user 0.86s system 99% cpu 7.211 total
⇒  time CACHE=1 node benchmark.js
842.385424000001
CACHE=1 node benchmark.js  2.10s user 0.82s system 101% cpu 2.885 total
⇒  time CACHE=1 node benchmark.js
819.7484239999997
CACHE=1 node benchmark.js  2.02s user 0.79s system 101% cpu 2.757 total
⇒  time CACHE=1 node benchmark.js
835.9143549999982
CACHE=1 node benchmark.js  2.08s user 0.82s system 101% cpu 2.868 total
⇒
````

File: lib/sync.js (hunk 1 above)
Comment (wonnage) [reply lib/sync.js]: Hmm, I could pull this logic into a separate module, we basically just need a memoized version of JSON.parse.

File: lib/sync.js [hunk 2]
```diff
@@ -101,6 +105,20 @@ module.exports = function (x, options) {
         }
     }
 
+    function parsePkg1(body) {
+        var key = crypto.createHash('md5').update(body).digest('hex');
```
Comment (lencioni) [inline lib/sync.js]: I think md4 would be sufficient here, and also be faster.

**Comment by wonnage** [comment]
I ran some more tests using the Benchmark npm module:

https://gist.github.com/wonnage/e33c9cb1fe9587236eb1ec11afd43772

````
uncached x 1,142 ops/sec ±5.57% (80 runs sampled)
cached x 4,022 ops/sec ±3.03% (83 runs sampled)
Fastest is cached
````

**Comment by ljharb** [comment]
ok, so in that example it's 4x faster - whether it's 4x or 6x, what savings does this amount to in real time?

**Comment by ljharb** [comment]
This seems like the sort of thing that might be better handled by a memoizing wrapper package (that a bundler could transparently replace `resolve` with). Closing for now.

**Pull request closed without merging**